feat: 템플릿 관련 추가/수정/삭제 로직 구현

wwwlnyy opened the pull request:
## 1️⃣ 어떤 작업을 했나요? (Summary)

템플릿 페이지에서 템플릿 조회, 추가, 수정, 삭제 기능을 구현했어요.

### templates/schema.ts
템플릿 api를 위한 스키마 정의

TemplateListResponse: 목록 조회 API의 전체 응답 타입

### templateOptions()
템플릿 쿼리 정의
- `templateKeys`: all, detail
- `templateOptions.all()`: 템플릿 전체 목록 조회
```typescript
all: () =>
    queryOptions({
      queryKey: templateKeys.all,
      queryFn: async () => {
        const data = await api.get('api/mails/templates').json();
        const response = TemplateListResponseSchema.parse(data).content;
        const templateData = response.map((template) => ({
          // 백엔드 응답을 프론트엔드 형식에 맞게 변환
          id: template.id,
          title: template.title,
          date: formatTemplates['2025.01.01'](template.updatedAt),
        }));
        return templateData;
      },
      staleTime: Infinity,
    }),
```
- `templateOptions.detail(templateId)`: 템플릿 상세 조회
```typescript
detail: (templateId: number) =>
    queryOptions({
      queryKey: templateKeys.detail(templateId),
      queryFn: async () => {
        const data = await api.get(`api/mails/templates/${templateId}`).json();
        const response = BaseTemplateSchema.parse(data);

        // 백엔드 응답 중 변수를 프론트엔드 형식에 맞게 변환
        const templateVariables = response.variables.map((variable) => {
          // 고정 변수(지원자명, 파트명)는 type이 null로 오기 때문에 미리 정의된 defaultVariables에서 찾아서 매핑
          if (!variable.type) {
            return defaultVariables.find((v) => v.id === variable.key)!;
          }
          // 그 외 변수들은 TypeMap을 사용하여 매핑
          // 나머지 항목들도 프론트엔드 형식에 맞게 변환
          return {
            id: variable.key,
            name: variable.displayName,
            type: TypeMap[variable.type],
            isFixed: false,
            differentForEachPerson: variable.perRecipient,
          };
        });

        // 최종적으로 템플릿 상세 데이터를 프론트엔드 형식에 맞게 변환
        const templateDetail = {
          id: response.id,
          title: response.title,
          content: transformBodyHtmlToContent(response.bodyHtml, templateVariables),
          variables: templateVariables,
          date: formatTemplates['2025.01.01'](response.updatedAt),
        };
        return templateDetail;
      },
      staleTime: Infinity,
    }),
```

### postTemplateFromForms()
템플릿 생성
```typescript
export const postTemplateFromForms = (params: PostTemplateParams) => {
  const formattedParams = {
    title: params.title,
    bodyHtml: transformContentToBodyHtml(params.content),
    variables: transformVariables(params.variables),
  };
  return api.post('api/mails/templates', { json: formattedParams }).json();
};
```

### deleteTemplate()
템플릿 삭제
```typescript
export const deleteTemplate = (params: DeleteTemplateParams) => {
  const { templateId } = params;
  return api.delete(`api/mails/templates/${templateId}`).json();
};
```

### putTemplate()
템플릿 수정
```typescript
export const putTemplate = (params: PutTemplateParams) => {
  const formattedParams = {
    title: params.title,
    bodyHtml: transformContentToBodyHtml(params.content),
    variables: transformVariables(params.variables),
  };
  return api.put(`api/mails/templates/${params.id}`, { json: formattedParams }).json();
};
```

## 2️⃣ 알아두시면 좋아요!
백엔드 수정사항 2가지가 반영이 되지 않아서 draft로 열어뒀어요.
1. 템플릿 수정 시 본문 길어지면 500에러 발생
2. 선언만 되고 사용하지 않은 변수 에러 처리 → 저장되도록 변경(어포던스)

지금까지 코드에서 크게 변하지는 않을 것 같아서 지금 리뷰 주셔도 될 것 같습니다.

## 3️⃣ 체크리스트 (Checklist)

- [x] `main` 브랜치의 최신 코드를 `pull` 받았나요?

wwwlnyy committed:
feat: 템플릿 관련 스키마 및 쿼리 옵션 정의

wwwlnyy committed:
refactor: Variable id 양식 변경

wwwlnyy committed:
refactor: VariableChipNode에서 key 속성 추가

wwwlnyy committed:
feat: templateOptions에 템플릿 상세 조회 로직 추가

wwwlnyy committed:
feat: 템플릿 생성 로직 구현

wwwlnyy committed:
refactor: template schema 변경

wwwlnyy committed:
feat: 백-프론트 응답 양식 맞추는 보조 함수 추가

wwwlnyy committed:
refactor: 템플릿 조회/생성 양식에 맞게 변경

wwwlnyy committed:
refactor: 템플릿 목록 화면에서 마지막 수정 시간 문구 변경

wwwlnyy committed:
feat: 템플릿 삭제 기능 구현

wwwlnyy committed:
feat: 템플릿 상세 조회 기능 구현

wwwlnyy committed:
feat: 템플릿 추가/상세조회/삭제 기능 적용

wwwlnyy committed:
fix: 빌드 에러 수정

wwwlnyy committed:
feat: 템플릿 수정 기능 구현

wwwlnyy committed:
refactor: 변수칩 뒤 공백 추가

wwwlnyy committed:
feat: 구글문서에서 복사-붙여넣기 시 span 정보 삭제

wwwlnyy committed:
refactor: template query key 수정

wwwlnyy committed:
refactor: tiptap 에디터에서 줄바꿈 속성 추가

wwwlnyy committed:
refactor: 에디터 내에서 복사 붙여넣기 시 변수 칩도 같이 복사되도록 수정

fecapark commented on src/pages/Template/components/TemplateEditor.tsx:
```diff
@@ -27,13 +27,13 @@ export const TemplateEditor = ({
     if (editorRef.current) {
       const chipType = getChipType(variable.type);
-      editorRef.current.insertVariable(chipType, variable.name);
+      editorRef.current.insertVariable(variable.id, chipType, variable.name);
```
이거 `deleteVariable` 은 label 기준으로 찾아서 제거하는데, 
insert 시점에 제공되는 key로 제거하지 않는 이유와, 그렇게 하지 않아도 괜찮을 이유가 있을까요?

fecapark commented on src/pages/Template/components/TemplateEditor.tsx:
```diff
@@ -27,13 +27,13 @@ export const TemplateEditor = ({
   const handleVariableAdd = (type: VariableType, name: string, differentForEachPerson: boolean) => {
     const newVariable: Variable = {
-      id: Date.now().toString(),
+      id: `var-${Date.now().toString()}`,
```
요건 무슨 이유로 추가된걸까요? 아무런 차이가 없지 않나요?

wwwlnyy replied:
1. defaultVariable에 uuid 반영: [86820d3](https://github.com/yourssu/Yourssu-Scouter-Frontend/pull/59/commits/86820d327dae84eb9d2c17d18880eb3d18e16d64)
2. defaultVariable > getDefaultVariable로 변경해 각 템플릿마다 다른 uuid를 갖도록 함: [b86abc2](https://github.com/yourssu/Yourssu-Scouter-Frontend/pull/59/commits/b86abc2caef388b86355882189434d197866e250)

fecapark commented on src/types/editor.ts:
```diff
@@ -9,6 +9,18 @@ export interface Variable {
+    isFixed: true,
+  },
+  {
+    id: 'applicant',
```
이런거보면 맥락이 얕은 사람들은 이유가 있을거라고 생각하게 되기 마련이라, 

`crypto.randomUUID` 를 써도 괜찮아보이네요.

fecapark replied:
어떤거는 고정키고 어떤거는 가변키인 것처럼 이분화될 필요가 있을까요?
문제가 없다면 이건 var- 로 맥락 통일하는게 DX 측면에서 좋아보여서요.
백엔드에 의존성이 있는 작업이라면 통일을 위한 소통을 해보시는게 좋을듯합니다

fecapark commented on src/utils/transformTemplate.ts:
```diff
@@ -0,0 +1,37 @@
+};
+
+export const transformBodyHtml = (content: string) => {
+  const regex = /<span[^>]*data-key="([^"]*)"[^>]*><\/span>/g;
```
1. 요거는 뭐하는 정규식인가요? span 태그이면서 data-key 어트리뷰트가 존재하는 모든 문자열인가요?
2. 정규식은 특히 변수명에 신경써주세요. 어떤 정규식인지 바로 이해할 수 있게요. `spanWithDataKeyAttribute`?

fecapark commented on src/query/template/schema.ts:
```diff
@@ -7,25 +7,34 @@ const VariableSchema = z.object({
+export type Template = z.infer<typeof BaseTemplateSchema>;
+export type TemplateListItem = z.infer<typeof TemplateListItemSchema>;
+export type TemplatePayload = z.infer<typeof TemplatePayloadSchema>;
+export type TemplateListResponse = z.infer<typeof TemplateListResponseSchema>;
```
저번에도 말씀드렸듯이, 이런 류의 커밋은 리뷰량을 기하급수적으로 늘리는 커밋 종류에요.

작업 진행도중 이전 커밋 내용에서 수정이 필요한 경우,
```
"아 이거 놓쳤네?" -> "refactor로 수정 커밋 추가해야겠다"
```
의 작업 흐름 보다는 

```
"아 이거 놓쳤네?" -> "rebase + fixup으로 커밋 정리 해야겠다"
```
의 흐름이 커밋량도 줄이고, 리뷰어가 가져야하는 컨텍스트량도 줄어들어요.

---

즉, 리뷰어를 위해서 https://github.com/yourssu/Yourssu-Scouter-Frontend/pull/59/commits/29b50fbe2676d2935a997f37591154594fc446b4 이런 커밋에 fixup 해주시면 좋아요.

항상 리뷰이도 리뷰어 입장에서 어떻게하면 작업대비 컨텍스트 량을 최소화할 수 있을지 
고민해보면 더 좋은 협업 실력을 기를 수 있을거에요.

wwwlnyy replied:
넵!! 명심하겠습니다

fecapark commented on src/pages/Template/components/EditTemplateDialog/EditTemplateDialog.tsx:
```diff
@@ -20,36 +22,38 @@ interface EditTemplateDialogProps {
+        title: templateDetail.title,
+        content: templateDetail.content,
+        variables: templateDetail.variables,
       });
```
이거 이제봤는데 templateDetail이 nullish하게 되는 경우는 언제인가요?
타입상에도 nullish 하지는 않아서요.

wwwlnyy replied:
음 없는 것 같네요... 수정했습니다
[refactor: 불필요한 nullish 검사 삭제](https://github.com/yourssu/Yourssu-Scouter-Frontend/commit/b58bca1c859f15f5ac04288192d0db8a2c4d205e)

wwwlnyy replied:
아뇨... 다시 했습니다...
[refactor: 템플릿 편집 내용 저장 시 불필요한 nullish 검사 삭제](https://github.com/yourssu/Yourssu-Scouter-Frontend/commit/8f252e70f5b4109a350f4ef5c16db04d6a225512)

fecapark commented on src/pages/Template/Templates.tsx:
```diff
@@ -68,6 +69,13 @@ export const Templates = () => {
+  const putTemplateMutation = useMutation({
+    mutationFn: putTemplate,
+    onSuccess: () => {
+      queryClient.invalidateQueries({ queryKey: ['templates'] });
```
src/query/template/options.ts에 있는 templateKeys.all 을 사용해주세요.

fecapark commented on src/query/template/mutations/putTemplate.ts:
```diff
@@ -0,0 +1,20 @@
+    variables: transformVariables(params.variables),
+  };
+
+  return api.put(`api/mails/templates/${params.id}`, { json: formattedParams }).json();
```
상동 (리턴 타입 정의하기 or 스키마 벨리데이션)

fecapark replied:
https://github.com/yourssu/Yourssu-Scouter-Frontend/pull/59/commits/f15b6f3bf21d2b5d0c965fbd4dd59ab53c9d1cbb 

여기서 됐네용 확인했습니다

fecapark commented on src/pages/Template/Templates.tsx:
```diff
@@ -109,22 +117,21 @@ export const Templates = () => {
   const handleSaveAdd = (newTemplate: Omit<Template, 'date' | 'id'>) => {
-    // 실제 추가 API 호출
-    console.log('템플릿 추가', newTemplate);
+    // console.log('템플릿 추가', newTemplate);
```
'안쓰는' 주석은 제거해주세요. 오히려 협업자의 맥락을 늘리고 코드가 더러워지는 등, 주석으로서의 존재 목적을 해치는 주범이에요

fecapark commented on src/pages/SendMail/MailEditorContent/MailEditorContent.style.ts:
```diff
@@ -14,6 +14,7 @@ export const StyledEditorContent = styled(EditorContent)`
   position: relative;
 
   .tiptap {
+    white-space: pre-wrap;
```
지금까지는 \n 줄바꿈이 안되었나요?

fecapark commented on src/pages/SendMail/MailEditorContent/MailEditorContent.tsx:
```diff
@@ -73,9 +73,11 @@ export const MailEditorContent = forwardRef<MailEditorContentRef, MailEditorCont
 
           doc.querySelectorAll('span').forEach((span) => {
+            if (span.hasAttribute('data-variable-chip')) {
+              return;
```
토막 상식) 

`hasAttribute` 메서드와 dataset 멤버를 통해 검증하는 방법의 성능 차이

https://www.measurethat.net/Benchmarks/Show/32279/0/getattribute-vs-dataset-without-explicit-loop

fecapark reviewed (commented):
고생하셨습니다.

좋았던 포인트) 
- 커밋마다 크기가 작아서 이해하는데 편했어요.

쓴소리) 
- 리뷰 과정에서 PR 디스크립션이 크게 도움되지는 않았어요. 어떻게 하면 작업의 맥락을 남길 수 있을지 고민해보시면 좋을거 같아요. 제가 작성했던 것들 참고해도 좋아요.

wwwlnyy committed:
fix: deleteVariable을 key(id) 기준으로 찾아서 제거하도록 변경

wwwlnyy committed:
fix: variables의 key 생성 방식을 Date에서 uuid로 변경

wwwlnyy committed:
refactor: 정규식 변수명 변경 및 주석 추가

wwwlnyy committed:
refactor: 템플릿 생성/수정 api의 response type 명시

wwwlnyy committed:
refactor: 날짜 포맷팅 함수를 formatKo로 통일

wwwlnyy committed:
refactor: 변수 타입 매핑 구조 개선

wwwlnyy committed:
refactor: 템플릿의 date 양식을 쿼리 레벨이 아닌 UI 레벨에서 resolve하도록 변경

wwwlnyy committed:
refactor: 불필요한 nullish 검사 삭제

wwwlnyy committed:
package: @tiptap/extension-placeholder

wwwlnyy committed:
feat: 메일 에디터 Placeholder 적용 및 빈 값 저장 처리

- Placeholder를 적용해 기본 문구를 대체하고 스타일링도 추가

- onCreate 핸들러를 추가해 빈 본문도 <p></p> 형태로 저장되도록 함

wwwlnyy committed:
chore: 주석 정리

wwwlnyy committed:
feat: 에디터 붙여넣기 스타일 제한 해제

- editorProps의 transformPastedHTML 로직 삭제

구글 문서 등에서 복사 시 원본 스타일이 유지되도록 변경

fecapark commented on src/query/template/mutations/postTemplatesFromForms.ts:
```diff
@@ -0,0 +1,19 @@
+    variables: transformVariables(params.variables),
+  };
+
+  return api.post('api/mails/templates', { json: formattedParams }).json();
```
api 응답이 템플릿 콘텐츠 타입 같은데, 식별자는 다르지만 같은 맥락인 `TemplateListItem` 타입으로 해결할 수 있지 않을까요?
반영할 계획이라면 이름도 TemplateItem 같은 이름으로 바꾸는게 좋을수도 있겠네요.

fecapark commented on src/pages/Template/Templates.tsx:
```diff
@@ -58,21 +58,30 @@ export const Templates = () => {
+          ...(oldData || {}),
+          ...updatedPayload,
+        };
+      });
```
요 부분을 결정하려면 

- tanstack query가 왜 '서버 상태 관리' 라이브러리라고 불리는지와, 
- 서버 상태 관리가 왜 필요한지,
- setQueryData와 invalidateQueries의 차이가 뭔지,

를 찾아보고 궁극적으로

> 뮤테이션 이후에 서버가 가지고 있는 정보와 클라이언트가 가지고 있는 정보가 불일치해도 상관없는가?
> (사용자 경험적 측면과 기술적 측면)

에 대해서 직접 고민해보는 시간을 가지면 좋을거 같아요.
나리 나름대로의 결론이 나면 여기에 적어주세요!

wwwlnyy committed:
refactor: Variable 인터페이스를 백엔드 스키마와 통일 및 키 생성 포맷 표준화

- 일관되지 않았던 가변 변수 키 생성 로직을  형태로 통일

wwwlnyy committed:
refactor: 템플릿 생성/수정 api의 response type을 TemplateItem으로 변경

wwwlnyy committed:
refator: 변수 타입 식별자 변경 및 transformVariables 함수 개선

wwwlnyy committed:
refactor: 템플릿 편집 내용 저장 시 불필요한 nullish 검사 삭제

wwwlnyy committed:
refactor: 템플릿 수정 후 캐시 데이터 동기화 개선

-서버 응답을 활용해 updatedAt이 즉시 갱신되도록 수정

wwwlnyy committed:
refactor: variable의 key, type 양식 변경

wwwlnyy committed:
refactor: variable > isFixed 필드를 삭제하고 type으로 구분

wwwlnyy committed:
refactor: defaultVariable > getDefaultVariable로 변경
